default.nix

```diff
@@ -117,34 +124,41 @@ with stdenv.lib; let
 in rec {
   mkNodeModules = { src, extraEnvVars ? {} }:
     let
+      localDeps = map dirOfLocal (builtins.filter isLocal (builtins.attrValues (lock.dependencies)));
       # filter out everything except package.json and package-lock.json if possible
       # allows to avoid rebuilding node_modules if these two files didn't change
       filteredSrc =
         let
           origSrc = if src ? _isLibCleanSourceWith then src.origSrc else src;
           getRelativePath = path: removePrefix (toString origSrc + "/") path;
-          usedPaths = [ "package.json" "package-lock.json" ];
+          usedPaths = [ "package.json" "package-lock.json" ]
+            ++ localDeps;
+          dirOfRec = x: if dirOf x == "." || dirOf x == "/" then [x] else ([x] ++ dirOfRec (dirOf x));
+          usedPathsRec = builtins.concatMap dirOfRec usedPaths;
           cleanedSource = cleanSourceWith {
             src = src;
-            filter = path: type: elem (getRelativePath path) usedPaths;
+            filter = path: type: any (x: elem x usedPathsRec) (dirOfRec (getRelativePath path));
           };
         in if canCleanSource src then cleanedSource else src;
 
-      packageJson = filteredSrc + "/package.json";
-      packageLockJson = filteredSrc + "/package-lock.json";
+
+      packageJson = src + "/package.json";
+      packageLockJson = src + "/package-lock.json";
```

Comment:
This path will change every time anything in `src` changes, causing everything to rebuild. As I understand the problem is that now `filteredSrc` depends on the contents of `package-lock.json`, causing infinite recursion. In that case we can create another filter expression which contains only `package.json` and `package-lock.json`, read `package-lock.json` from there, and then use it to construct a filter which will include local dependencies

Reply:
Oh, good point.

Reply:
@balsoft Do you plan to address this in this PR or would you like to create a separate issue?

Reply:
I think it's a separate issue.

Reply:
Why not do it in this PR?

Reply:
Maybe fixed it here, please take a look

Reply:
> Why not do it in this PR?

_Because it has the potential for postponing the merging of this PR forever._

Reply:
In my opinion caching node_modules is more important than `file:` dependency support. @balsoft I can help you with finishing this PR if you want

Reply:
There're only minor fixes left as I see